ntp-proto/src/filter.rs

```diff
@@ -359,28 +357,56 @@ fn construct_candidate_list<'a>(
     candidate_list
 }
 
-/// Find the largest contiguous intersection of correctness
-/// intervals.  Allow is the number of allowed falsetickers;
-/// found is the number of midpoints.  Note that the edge values
-/// are limited to the range +-(2 ^ 30) < +-2e9 by the timestamp
-/// calculations.
 #[allow(dead_code)]
-fn find_interval(chime_list: &[CandidateTuple]) -> (NtpDuration, NtpDuration) {
+struct SurvivorTuple<'a> {
+    p: &'a Peer,
+    metric: NtpDuration,
+}
+
+/// Collect the candidates within the correctness interval
+#[allow(dead_code)]
+fn construct_survivors<'a>(
+    chime_list: &[CandidateTuple<'a>],
+    local_clock_time: NtpTimestamp,
+) -> Vec<SurvivorTuple<'a>> {
+    let mut survivors = Vec::new();
+
+    if let Some((low, high)) = find_interval(chime_list) {
```

Comment:
You could do an early return here, which I think would make this code more readable.

It seems like this function (and potentially other functions in this module) are filtering (and enriching?) an initial list. It might be nice to avoid allocations while doing so and instead spin this as a `filter_survivor(candidate: CandidateTuple<'a>, local_clock_time: NtpTimestamp) -> Option<SurvivorTuple<'a>>`.

Reply:
yes that worked out nicely!